src/page/lru.rs

```diff
@@ -18,139 +15,130 @@ impl Lru {
         let size = 2 << cache_bits;
+        let shard_capacity = cache_capacity / size;
         Lru {
-            shards: rep_no_copy![RwLock::new(Shard::default()); size],
-            cache_capacity: cache_capacity,
+            shards: rep_no_copy![Mutex::new(Shard::new(shard_capacity)); size],
             cache_bits: cache_bits,
-            approx_size: AtomicIsize::new(1),
         }
     }
 
-    /// Whenever data is paged-in, we set up bookkeeping for its accesses here.
-    /// Returns a Vec of items that should now be paged-out to stay under the
-    /// configured cache size.
-    pub fn paged_in(&self, pid: PageID, lid: LogID, sz: usize) -> Vec<(PageID, LogID)> {
-        let k: [u8; 8] = unsafe { std::mem::transmute(lid) };
-        let h = hash::hash(&k, 0) as usize;
-        let idx = h % (2 << self.cache_bits);
-
-        let entry = Entry {
-            pid: pid,
-            lid: lid,
-            mtime: time::get_time().sec,
-            sz: sz,
-            accesses: Rc::new(AtomicUsize::new(1)),
-        };
-
+    /// Called when a page is accessed. Returns a Vec of pages to
+    /// try to page-out. For each one of these, the caller is expected
+    /// to call `page_out_succeeded` if the page-out succeeded.
+    pub fn accessed(&self, pid: PageID, sz: usize) -> Vec<PageID> {
+        let idx = self.idx(pid);
         let ref shard_mu = self.shards[idx];
-        let mut shard = shard_mu.write().unwrap();
-        shard.insert(entry);
-
-        let mut to_evict = vec![];
-        let mut cur_sz = self.approx_size.fetch_add(sz as isize, Ordering::SeqCst) + sz as isize;
-        while cur_sz > self.cache_capacity as isize {
-            if shard.len() == 1 {
-                // don't evict what we just added
-                break;
-            }
-            let min = shard.min();
-            shard.pop(min.lid).unwrap();
-            to_evict.push((min.pid, min.lid));
-            cur_sz -= min.sz as isize;
-            self.approx_size.fetch_sub(min.sz as isize, Ordering::SeqCst);
-        }
-
-        to_evict
+        let mut shard = shard_mu.lock().unwrap();
+        shard.accessed(pid, sz)
     }
 
-    /// Bumps the LRU stats for this page.
-    pub fn accessed(&self, lid: LogID) {
-        let k: [u8; 8] = unsafe { std::mem::transmute(lid) };
-        let h = hash::hash(&k, 0) as usize;
-        let idx = h % (2 << self.cache_bits);
-
+    /// Signal that a page-out was successful.
+    pub fn page_out_succeeded(&self, pid: PageID) {
+        let idx = self.idx(pid);
         let ref shard_mu = self.shards[idx];
-        let mut shard = shard_mu.write().unwrap();
+        let mut shard = shard_mu.lock().unwrap();
+        shard.page_out_succeeded(pid);
+    }
 
-        shard.bump(lid);
+    fn idx(&self, pid: PageID) -> usize {
+        let k: [u8; 8] = unsafe { std::mem::transmute(pid) };
+        let h = hash::hash(&k, 0) as usize;
+        h % (2 << self.cache_bits)
```

Comment:
this function turns an 8 byte long `PageID` into an array of 8 bytes, then passes that array to the hash function, then uses modulus on the cache size (we have 2 ^ cache_bits total shards).